### PR TITLE
Upgrade to React 17

### LINKED ISSUE #6910: Upgrade to React 17
Currently, all backstage packages depend on React 16. React 17 has been stable for quite a while now, and provides better handling of multiple React versions running concurrently. Eventually using the new jsx factory would be great, but let this issue be only about bumping everything to 17.

Since this is a breaking change, it should be done in a breaking version bump.

I'm not 100% sure of what else will be needed for this, but at least `react` and `react-dom` need to be upgraded together.

On current master it's mostly 16.13.1, but a few other have sneaked in:

```
❯ rg --files -g package.json | xargs jq '.dependencies.react' | grep -v null | sort | uniq -c | sort -rn
  43 "^16.13.1"
   8 "^16.12.0"
   1 "^17.0.2"
   1 "^16.8.5"
   1 "^16.0.0"
```

When testing locally, I found that pinning `react` and `react-dom` to `17.0.2` and `@hot-loader/react-dom` to `17.0.1` works. Haven't done any thorough testing, but the UI loads and navigation works fine from what I can see.

Do we know of any other blocking issues before bumping?

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

Fixes #6910

All plugins used in the main repo now declare a peer dependency on react, with the version range "16.x.x OR 17.x.x". This means we simply switch out the `react` and `react-dom` dependency of the app, and we end up on 17. Still provided instructions for how to use yarn resolutions in the changeset though.

Verified by browsing around a bit, but in the end this is ofc controlled by each app so nobody's forced to move if it causes trouble. We've simply enabled the migration and it can done later if needed.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by benjdlambert — APPROVED** [review]
Wooooo

**Comment by Rugvip** [comment]
When I said tested I mean "forgot that I had checked that rolling back to v16 works and then did all the testing there", looking into what's up :grin:

**Comment by Rugvip** [comment]
Alright, fixed and properly tested. Also made sure hot reloading works

File: packages/cli/src/lib/bundler/config.ts [hunk 1]
```diff
@@ -173,6 +173,18 @@ export async function createConfig(
     }),
   );
 
+  const resolveAliases: Record<string, string> = {};
+  try {
+    // eslint-disable-next-line import/no-extraneous-dependencies
+    const { version: reactDomVersion } = require('react-dom/package.json');
+    // Only apply the alias for hook support if we're running with React 16
+    if (reactDomVersion.startsWith('16.')) {
```
Comment (freben) [inline packages/cli/src/lib/bundler/config.ts:181]: it's good that we have the dot here, so that we don't get a version 160 problem down the line

**Review by freben — APPROVED** [review]
💛

**Review by jhaals — APPROVED** [review]
Great 👏👏

**Pull request merged**